sycl/doc/extensions/Assert/SYCL_ONEAPI_ASSERT.asciidoc

```diff
@@ -0,0 +1,131 @@
+
+// This is necessary for asciidoc, but not for asciidoctor
+:cpp: C++
+
+== Notice
+IMPORTANT: This specification is a draft.
+
+Copyright (c) 2021-2021 Intel Corporation. All rights reserved.
+
+NOTE: Khronos(R) is a registered trademark and SYCL(TM) and SPIR(TM) are
+trademarks of The Khronos Group Inc.  OpenCL(TM) is a trademark of Apple Inc.
+used by permission by Khronos.
+
+NOTE: This document is better viewed when rendered as html with asciidoctor.
+GitHub does not render image icons.
+
+== Dependencies
+
+This extension is written against the SYCL 2020 specification, Revision 3.
+
+== Status
+
+Working Draft
+
+This is a preview extension specification, intended to provide early access to
+a feature for review and community feedback. When the feature matures, this
+specification may be released as a formal extension.
+
+Because the interfaces defined by this specification are not final and are
+subject to change they are not intended to be used by shipping software
+products.
+
+== Introduction
+This extension adds the ability for device code to call the C++ `assert()`
+macro. The behavior of `assert()` in device code is similar to its behavior in
+host code. If the asserted condition is false, a message is printed to `stderr`
+and then the program typically aborts. The mechanism for aborting the program is
+different, though. Whereas the host version calls `std::abort()`, the device
+version causes an asynchronous SYCL `exception` with the
+`errc::ext_oneapi_assert` error code to be thrown. The application can catch
+this exception like any other asynchronous exception that is thrown from an
+executing kernel. The numeric value of this enumeration element is defined by
+implementation.
+
+
+Catching the asynchronous exception with `sycl::errc::ext_oneapi_assert` error
+code means that assert failure had happened in device code during kernel
+execution and the assert message is already printed to `stderr`. Format of the
+assert message is the following:
+```
+<file>:<line>: <function>: global id: [<gid0>,<gid1>,<gid2>], local id: [<lid0>,<lid1>,<lid2>] Assertion `<expr>` failed.
+```
+
+It is unspecified whether a failing `assert()` returns to its caller before the
+kernel terminates. If a failing call returns, the device code may need to
+continue execution without deadlocking for the assertion message to be printed
+or for the exception to be thrown.
+
+The contents of the exception's `e.what()` string is unspecified. Since the
+assertion message is already printed to `stderr` by the time the exception is
+thrown, the `e.what()` string may not include the location of the failing
+assertion.
+
+The `assert()` macro is defined in system include headers, not in SYCL headers.
+On most of systems it is `<cassert>` and/or `<assert.h>` header files.
+The user can can disable assertions in device code by defining the `NDEBUG`
+preprocessor macro at the point when `<sycl.hpp>` and `<cassert>/<assert.h>`
+are included.
```

Comment:
These are two different points in the application.  If the application wants to disable assertions, should it define `NDEBUG` before including `<sycl.hpp>` or before including `<cassert>/<assert.h>`?  Or, maybe it needs to be defined before including either?  If that is the case, you should say that.

This is an important detail because some applications will put `#define NDEBUG` in their source code, and those applications need to know where to put this `#define`.

Reply:
Fixed.